File: 0055-jump-game/solution.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        int maxReach = 0;

        /*maxReach = farthest index reachable so far
        If you ever reach an index you can’t get to, the answer is false
        */

        for(int i=0; i<n; i++) {
            if(i > maxReach) {
                return false;
            }
            maxReach = Math.max(maxReach, i+nums[i]);
        }
        return true;
    }
}

File: 0055-jump-game/solution.swift
class Solution {
    func canJump(_ nums: [Int]) -> Bool {
        var maxReach = 0
        for i in 0..<nums.count {
            if i <= maxReach {
                maxReach =  max(maxReach, i + nums[i])
            }
        }
        return maxReach >= nums.count - 1
    }
}
